### DISCUSSION TITLE
Auto resize window

### DISCUSSION THREAD

**Discussion by frwol** [TomSchimansky/CustomTkinter#435] [Q&A] [answered]
Hi,

I'm trying to create a dynamic GUI where a variable number of labels, entries and buttons are created based on the content of a specific file. Everything works great except that the main window doesn't auto resize and keeps a fixed size (which seems to be hard coded to 500x600). My window has two frames : one with just a label + entry + button and another one with several rows of  label + entry + button. All widgets are placed using the .grid() function. The second frame correctly resize itself based on the number of widgets that are created but not the main window.
I'm sure I'm missing something obvious but I can't tell what.

FYI I'm using customtkinter 4.5.10 and Python 3.10.6.

**Comment by frwol** [comment]
Ok I've done some tests and tried with just a really simple application with just one frame and several buttons. I've tried the same layout, same code (nothing less, nothing more) but with Tkinter, and in this case the window properly resizes itself. So it really is related to CTk.

```python
import tkinter
import customtkinter

class TestAppCTk():
    def __init__(self) -> None:
        self._window = customtkinter.CTk()

        frame = customtkinter.CTkFrame(master=self._window)
        frame.grid(row=0, column=0, padx=15, pady=15)

        # Create 20 buttons
        for btn_idx in range(20):
            btn = customtkinter.CTkButton(master=frame)
            btn.grid(row=btn_idx, column=0, padx=15, pady=5)
            btn.set_text(f"Button {btn_idx}")

class TestAppTk():
    def __init__(self):
        self._window = tkinter.Tk()

        # Create a frame
        self._frame = tkinter.Frame(master=self._window)
        self._frame.grid(row=0, column=0, padx=15, pady=15)

        # Create 20 buttons
        for btn_idx in range(20):
            btn = tkinter.Button(master=self._frame, text=f"Button {btn_idx}")
            btn.grid(row=btn_idx, column=0, padx=15, pady=5)

if __name__ == "__main__":
    app = TestAppCTk()
    app._window.mainloop()

    #app = TestAppTk()
    #app._window.mainloop()
```

**Comment by frwol** [comment]
Result when using Tkinter:
![Screenshot from 2022-09-08 13-28-19](https://user-images.githubusercontent.com/60254299/189111297-7c0b25c4-c727-4e24-a8bf-c78339ca73e1.png)

Result when using CustomTkinter
![Screenshot from 2022-09-08 13-28-37](https://user-images.githubusercontent.com/60254299/189111300-4bcaad1d-9732-4095-bfe1-713ca065faa8.png)

**Comment by Akascape** [comment, marked as answer]
@frwol I got the solution for you,

First make a separate **frame** which will act as the background window (do not use _pady_ or _padx_). 

``` python
self._window = customtkinter.CTk()
frame = customtkinter.CTkFrame(master=self._window)
frame.grid(row=0, column=0)
```

Create all of your widgets/sub-frames inside that main frame and then add these lines in the end of your code:

``` python
frame.update_idletasks()
width = frame.winfo_width()
height = frame.winfo_height()
self._window.geometry(f"{width}x{height}")
```

Here is your modified [example file](https://github.com/TomSchimansky/CustomTkinter/files/9542305/example_modified.zip)

![Example Modified Screenshot](https://user-images.githubusercontent.com/89206401/189523541-1e2c5bda-bdec-4622-af30-200e9dce4f30.png)

**Reply by frwol** [reply]
Hi @Akascape,
thanks for your answer. This is exactly what I was looking for ! Now I'm curious about the technical details. Why do I need to create a "main" frame and update the display to get the correct dimensions when using CustomTkinter and not when using Tkinter ?

**Reply by Akascape** [reply]
@frwol 
The default CTK window does not change the width/height on addition of any widgets, but a CTK frame adjust itself based on the widgets placed inside it. So we are using that frame as the main window.
I don't know why the CTK window is not adjusting automatically even if we use the same concept, maybe it's not implemented yet.

So that's why we have to use that frame height & width for the CTK window. We have used the `widget.winfo_width()` and `widget.winfo_height()` method because the normal `widget['height']` or `widget['width']` method doesn't give us the correct value.

In the end we have just updated the geometry for the CTK window using `window.geometry(f"{width}x{height}")`

That's all the explaination I can give for now. Thanks for asking...

**Reply by avalon60** [reply]
This has been a major cause of frustration for me. Hopefully will get fixed,

**Comment by NZ62AN** [comment]
very good.
